Enable ThinLTO for rustc on x64 msvc

lqd opened the pull request:
This applies the great work from @bjorn3 and @Kobzol in https://github.com/rust-lang/rust/pull/101403 to x64 msvc.

Here are the local results for the try build `68c5c85ed759334a11f0b0e586f5032a23f85ce4`, compared to its parent `0a6b941df354c59b546ec4c0d27f2b9b0cb1162c`. Looking better than my previous local builds.

![image](https://user-images.githubusercontent.com/247183/198158039-98ebac0e-da0e-462e-8162-95e88345edb9.png)

(I can't show cycle counts, as that option is failing on the windows version of the perf collector, but I'll try to analyze and debug this soon)

This will be the first of a few tests for rustc / llvm / both ThinLTO on the windows and mac targets.

lqd committed:
enable ThinLTO for rustc on x86_64-pc-windows-msvc dist builds

bors commented:
:hourglass: Trying commit 0abf95bdd7e7f98b53c310649ecb03bc60d986f6 with merge 00df0000c5e43d256786737e9f49379e1b3054c4...
<!-- homu: {"type":"TryBuildStarted","head_sha":"0abf95bdd7e7f98b53c310649ecb03bc60d986f6","merge_sha":"00df0000c5e43d256786737e9f49379e1b3054c4"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3331539764/jobs/5511410147)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/3331539764/jobs/5511410147","builder_name":"checks-actions"} -->

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/3331539764/jobs/5511396434) [(plain)](https://github.com/rust-lang-ci/rust/commit/00df0000c5e43d256786737e9f49379e1b3054c4/checks/9124513274/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Updating files:  99% (38115/38499)
Updating files:  99% (38423/38499)
Updating files: 100% (38499/38499)
Updating files: 100% (38499/38499), done.
branch 'try' set up to track 'origin/try'.
Switched to a new branch 'try'
[command]"C:\Program Files\Git\bin\git.exe" log -1 --format='%H'
'00df0000c5e43d256786737e9f49379e1b3054c4'
##[group]Run src/ci/scripts/setup-environment.sh
src/ci/scripts/setup-environment.sh
```

</details>

lqd commented:
@bors try

bors commented:
:hourglass: Trying commit 0e16a37a9f376091f4573feb33d0bcd603c811e8 with merge 68c5c85ed759334a11f0b0e586f5032a23f85ce4...
<!-- homu: {"type":"TryBuildStarted","head_sha":"0e16a37a9f376091f4573feb33d0bcd603c811e8","merge_sha":"68c5c85ed759334a11f0b0e586f5032a23f85ce4"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3331864728/jobs/5513925650)
Build commit: 68c5c85ed759334a11f0b0e586f5032a23f85ce4 (`68c5c85ed759334a11f0b0e586f5032a23f85ce4`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3331864728/jobs/5513925650"},"merge_sha":"68c5c85ed759334a11f0b0e586f5032a23f85ce4"} -->

lqd commented:
To compare with the size increase we see on linux:
- `rustc_driver.dll`: 3% increase (≈4MB), from 124 069 888 to 128 485 888
- `rustc_driver.pdb`: 5% increase (≈4MB), from 87 977 984 to 92 418 048

lqd commented:
I've updated the PR description with the results of the `try` build, and dropped the temporary CI hacks. 

I will look at the sizes of the `rustc-dev` component tomorrow, just in case we see an increase like we did on linux. Otherwise, I think this is ready to review, feel free to re-roll: 

r? @Mark-Simulacrum

lqd commented:
@rustbot ready

I don't have access to the MSVC builder times graph that t-infra has, but from the tests above, it seems the additional time wouldn't necessarily be impactful on CI times, and that we could land this change @Mark-Simulacrum ?

Mark-Simulacrum commented:
This is the current state of the world:

![image](https://user-images.githubusercontent.com/5047365/206327284-99b82f21-f1e8-4fb9-823b-2e4b6ec39d64.png)

So msvc dist is ~3rd slowest today. I think this should be OK to merge though, we can adjust further as we iterate on the CI times -- I expect our planned improvements around LLVM caching on Linux should fairly cleanly map directly to the msvc builder here.

@bors r+

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3644193663/jobs/6153370838)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/3644193663/jobs/6153370838","builder_name":"checks-actions"} -->

ChrisDenton commented:
@bors retry spurious toolstate failure (see [Zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/242791-t-infra/topic/Authentication.20failed.20on.20merge.20attempt/near/314590516))

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/3644193663/jobs/6153195664) [(plain)](https://github.com/rust-lang-ci/rust/commit/da2270c1464b4af6c57a1193a1600cd7dd352c98/checks/9956491037/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 * branch            master     -> FETCH_HEAD
HEAD is now at 496fb40 (windows CI update)
[master c75804d] (linux CI update)
 1 file changed, 1 insertion(+)
fatal: could not read Username for 'https://github.com': No such device or address
Sleeping for 3 seconds before retrying push
 * branch            master     -> FETCH_HEAD
HEAD is now at 496fb40 (windows CI update)
[master 918f39c] (linux CI update)
 1 file changed, 1 insertion(+)
 1 file changed, 1 insertion(+)
fatal: could not read Username for 'https://github.com': No such device or address
Sleeping for 3 seconds before retrying push
 * branch            master     -> FETCH_HEAD
HEAD is now at 496fb40 (windows CI update)
[master 9cf58d6] (linux CI update)
 1 file changed, 1 insertion(+)
 1 file changed, 1 insertion(+)
fatal: could not read Username for 'https://github.com': No such device or address
Sleeping for 3 seconds before retrying push
 * branch            master     -> FETCH_HEAD
HEAD is now at 496fb40 (windows CI update)
[master a961ffc] (linux CI update)
 1 file changed, 1 insertion(+)
 1 file changed, 1 insertion(+)
fatal: could not read Username for 'https://github.com': No such device or address
Sleeping for 3 seconds before retrying push
 * branch            master     -> FETCH_HEAD
 * branch            master     -> FETCH_HEAD
thread 'main' panicked at 'Failed to update toolstate repository with new data', toolstate.rs:436:9
HEAD is now at 496fb40 (windows CI update)
Build completed unsuccessfully in 0:00:17
```

</details>

bors commented:
:hourglass: Testing commit 684663ed380d0e6a6e135aed9c6055ab4ba94ac8 with merge 657eefe2dcf18f76ac67a39945810128e101178c...
<!-- homu: {"type":"BuildStarted","head_sha":"684663ed380d0e6a6e135aed9c6055ab4ba94ac8","merge_sha":"657eefe2dcf18f76ac67a39945810128e101178c"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3670078468/jobs/6205377071)
Approved by: Mark-Simulacrum
Pushing 657eefe2dcf18f76ac67a39945810128e101178c to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Mark-Simulacrum","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3670078468/jobs/6205377071"},"merge_sha":"657eefe2dcf18f76ac67a39945810128e101178c"} -->

rust-timer commented:
Finished benchmarking commit (657eefe2dcf18f76ac67a39945810128e101178c): [comparison URL](https://perf.rust-lang.org/compare.html?start=d137783642b0b98eda2796dc66bffc2b089a8327&end=657eefe2dcf18f76ac67a39945810128e101178c&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=d137783642b0b98eda2796dc66bffc2b089a8327&end=657eefe2dcf18f76ac67a39945810128e101178c&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=d137783642b0b98eda2796dc66bffc2b089a8327&end=657eefe2dcf18f76ac67a39945810128e101178c&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 3.0%  | [3.0%, 3.0%]   | 1     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -1.3% | [-2.0%, -0.7%] | 2     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=d137783642b0b98eda2796dc66bffc2b089a8327&end=657eefe2dcf18f76ac67a39945810128e101178c&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.1% | [-2.1%, -2.1%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

nnethercote commented:
Note: the perf CI result showed no change because it only measures on Windows. See the image at the top of this PR for some previously-measured Windows results.
